harness/determined/searcher/search_runner.py

```diff
@@ -0,0 +1,112 @@
+
+        # searcher_state = exp.get
+        assert client._determined is not None
+        session = client._determined._session
+        experiment_id = exp._id
+        logging.debug(f"Running experiment {experiment_id}")
+
+        experiment_is_active = True
+
+        try:
+            while experiment_is_active:
+                time.sleep(5)
+                events = bindings.get_GetSearcherEvents(session, experimentId=experiment_id)
+                if events.searcherEvents is None:
+                    continue
+                logging.warning(json.dumps([e.to_json() for e in events.searcherEvents]))
+                for e in events.searcherEvents:
+                    if e.initialOperations:
+                        logging.info("initial operations")
+                        operations = self.search_method.initial_operations()
+                    elif e.trialCreated:
+                        logging.info(f"trialCreated({e.trialCreated.requestId})")
+                        operations = self.search_method.on_trial_created(
+                            uuid.UUID(e.trialCreated.requestId)
+                        )
+                    elif e.trialClosed:
+                        logging.info(f"trialClosed({e.trialClosed.requestId})")
+                        operations = self.search_method.on_trial_closed(
+                            uuid.UUID(e.trialClosed.requestId)
+                        )
+                    elif e.trialExitedEarly:
+                        logging.info(
+                            f"trialExitedEarly({e.trialExitedEarly.requestId},"
+                            f" {e.trialExitedEarly.exitedReason})"
+                        )
+                        if e.trialExitedEarly.exitedReason is None:
+                            raise RuntimeError(
+                                "trialExitedEarly event is invalid without exitedReason"
+                            )
+                        operations = self.search_method.on_trial_exited_early(
+                            uuid.UUID(e.trialExitedEarly.requestId),
+                            exited_reason=ExitedReason._from_bindings(
+                                e.trialExitedEarly.exitedReason
+                            ),
+                        )
+                    elif e.validationCompleted:
+                        logging.info(
+                            f"validationCompleted({e.validationCompleted.requestId},"
+                            f" {e.validationCompleted.metric})"
+                        )
+                        if e.validationCompleted.metric is None:
+                            raise RuntimeError(
+                                "validationCompleted event is invalid without a metric"
+                            )
+                        operations = self.search_method.on_validation_completed(
+                            e.validationCompleted.metric
+                        )
+                    elif e.experimentInactive:
```

Comment:
what would generate the event master side?

Reply:
This relates to both of your questions.

Looking at the master logic, I see that Searcher.ValidationCompleted does take an operation, of `ValidateAfter` type. It is used to confirm that the SearchMethod is only notified once about the completion of `ValidateAfter` for a specific trial at a specific length. It is not passed on to the SearchMethod.
This event is generated here https://github.com/determined-ai/determined/blob/fb4dba510677e9a6472e92de2e2a4b7da98e516f/master/internal/api_trials.go#L901
and that is called from https://github.com/determined-ai/determined/blob/fb4dba510677e9a6472e92de2e2a4b7da98e516f/harness/determined/core/_searcher.py#L85

Reply:
i am asking about the `experimentInactive` event

Reply:
when client asks master for events, the first thing master does is check the status of the experiment. If the experiment is completed the master responds with the list containing a single event, and that is event is experimentInactive. I am not sure this exactly the right thing to do - maybe it is better to return 404 not found. We will figure it out outside this PR.